### PR TITLE
Dual-Button IME (on-device input method) for Trezor One

### PR DESCRIPTION
**Pull request by exeabyss**
Dual-Button IME support for passphrase input is in. The support for PIN input and seed recovery is to be done.

### CODE REVIEW

**Comment by tsusanka** [comment]
The tests are failing because of style. Please run `pipenv run make style` to fix that.

**Comment by tsusanka** [comment]
Or just apply this commit: https://github.com/trezor/trezor-firmware/commit/3054a562881257c394a0285b68c32d99069754de

---

Not: the GitLab build is running [here](https://gitlab.com/satoshilabs/trezor/trezor-firmware/pipelines/96608363)

**Comment by exeabyss** [comment]
The style is fixed.

**Comment by tsusanka** [comment]
Thank you for your contribution! I'm surprised how usable this actually is! We evaluate this and @prusnak will get back to you.

File: legacy/gen/font.inc [hunk 1]
```diff
@@ -1,3 +1,35 @@
...
+	/* 0x16 _ */ (uint8_t *)"\x01\x00",
+	/* 0x17 _ */ (uint8_t *)"\x01\x00",
+	/* 0x18 _ */ (uint8_t *)"\x01\x00",
+	/* 0x19 _ */ (uint8_t *)"\x01\x00",
+	/* 0x1a _ */ (uint8_t *)"\x01\x00",
+	/* 0x1b _ */ (uint8_t *)"\x01\x00",
+	/* 0x1c _ */ (uint8_t *)"\x01\x00",
+	/* 0x1d _ */ (uint8_t *)"\x01\x00",
+	/* 0x1e _ */ (uint8_t *)"\x01\x00",
+	/* 0x1f _ */ (uint8_t *)"\x01\x00",
```
Comment (prusnak) [inline legacy/gen/font.inc]: Please remove this change. It got here by accident (by wrong rebase I suppose).

File: legacy/gen/fontfixed.inc [hunk 2]
```diff
@@ -1,19 +1,51 @@
...
+	/* 0x16 _ */ (uint8_t *)"\x01\x00",
+	/* 0x17 _ */ (uint8_t *)"\x01\x00",
+	/* 0x18 _ */ (uint8_t *)"\x01\x00",
+	/* 0x19 _ */ (uint8_t *)"\x01\x00",
+	/* 0x1a _ */ (uint8_t *)"\x01\x00",
+	/* 0x1b _ */ (uint8_t *)"\x01\x00",
+	/* 0x1c _ */ (uint8_t *)"\x01\x00",
+	/* 0x1d _ */ (uint8_t *)"\x01\x00",
+	/* 0x1e _ */ (uint8_t *)"\x01\x00",
+	/* 0x1f _ */ (uint8_t *)"\x01\x00",
```
Comment (prusnak) [inline legacy/gen/fontfixed.inc]: Please remove this change. It got here by accident (by wrong rebase I suppose).

File: legacy/gen/fonts.c [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #include "fonts.h"
 
-const uint8_t *const font_data[FONTS][128 - 32] = {
+const uint8_t *const font_data[FONTS][128] = {
```
Comment (prusnak) [inline legacy/gen/fonts.c]: Please remove this change. It got here by accident (by wrong rebase I suppose).

File: legacy/gen/fonts.c [hunk 4]
```diff
@@ -12,10 +12,9 @@ const uint8_t *const font_data[FONTS][128 - 32] = {
 };
 
 int fontCharWidth(uint8_t font, uint8_t c) {
-  return (c < 0x20 || c >= 0x80) ? 0 : font_data[font % FONTS][c - 32][0];
+  return (c >= 0x80) ? 0 : font_data[font % FONTS][c][0];
```
Comment (prusnak) [inline legacy/gen/fonts.c]: Please remove this change. It got here by accident (by wrong rebase I suppose).

File: legacy/gen/fonts.c [hunk 5]
```diff
@@ -12,10 +12,9 @@ const uint8_t *const font_data[FONTS][128 - 32] = {
...
 }
 
 const uint8_t *fontCharData(uint8_t font, uint8_t c) {
-  return (c < 0x20 || c >= 0x80) ? (const uint8_t *)""
-                                 : font_data[font % FONTS][c - 32] + 1;
+  return (c >= 0x80) ? (const uint8_t *)"" : font_data[font % FONTS][c] + 1;
```
Comment (prusnak) [inline legacy/gen/fonts.c]: Please remove this change. It got here by accident (by wrong rebase I suppose).

File: legacy/gen/fonts.h [hunk 6]
```diff
@@ -15,7 +15,7 @@
 
 #define FONT_DOUBLE 0x80
 
-extern const uint8_t *const font_data[FONTS][128 - 32];
+extern const uint8_t *const font_data[FONTS][128];
```
Comment (prusnak) [inline legacy/gen/fonts.h]: Please remove this change. It got here by accident (by wrong rebase I suppose).

File: legacy/gen/fonts/generate.py [hunk 7]
```diff
@@ -23,7 +23,7 @@ def convert(imgfile, outfile):
     img = Img(imgfile)
     cur = ""
     with open(outfile, "w") as f:
-        for i in range(128 - 32):
+        for i in range(128):
```
Comment (prusnak) [inline legacy/gen/fonts/generate.py]: Please remove this change. It got here by accident (by wrong rebase I suppose).

File: legacy/gen/fonts/generate.py [hunk 8]
```diff
@@ -32,10 +32,14 @@ def convert(imgfile, outfile):
                 x += 1
                 cur += "\\x%02x" % int(val, 2)
             cur = "\\x%02x" % (len(cur) // 4) + cur
-            i += 32
-            ch = "_" if (i == 127) else chr(i)
+            ch = chr(i) if i >= 32 and i <= 126 else "_"
```
Comment (prusnak) [inline legacy/gen/fonts/generate.py]: Please remove this change. It got here by accident (by wrong rebase I suppose).

File: legacy/gen/fontfixed.inc [hunk 9]
```diff
@@ -1,19 +1,51 @@
...
 	/* 0x20   */ (uint8_t *)"\x01\x00",
-	/* 0x21 ! */ (uint8_t *)"\x03\x60\xfa\x60",
+	/* 0x21 ! */ (uint8_t *)"\x05\x00\x60\xfa\x60\x00",
```
Comment (prusnak) [inline legacy/gen/fontfixed.inc]: Can you please edit `fontfixed.png` accordingly? This file is generated from this image, so the png file should match.

**Comment by exeabyss** [comment]
The changes are necessary. Unprintable characters are used to display special characters used in Dual-Button IME, such as Space, Backspace, Done, and the caret.

fontfixed.png was edited. Besides that special characters were added, all printable characters were made the same width since the name suggested the font was fixed-width. generate.py and font.png were reverted to support all characters before 0x80.

**Comment by tsusanka** [comment]
Hi there, thanks a lot for your contribution. We are currently porting our "Core" codebase to T1 and we do not want to introduce large code changes in the current T1 code. We really like your PR, though, and we will consider implementing it to T1 again in Core (in the Python codebase). Thanks a lot!

**Pull request closed without merging**